src/v/archival/ntp_archiver_service.cc

```diff
@@ -532,8 +538,13 @@ bool ntp_archiver::housekeeping_can_continue() const {
 
 ss::future<> ntp_archiver::stop() {
     _leader_cond.broken();
+    if (_local_segment_merger) {
+        co_await dynamic_cast<local_adjacent_segment_merger*>(
```

Comment:
Coud we just use the derived type instead (i.e. change the type of `local_segment_merger` to `unique_ptr<local_adjacent_segment_merger>`)? I think that would remove the need for dynamic cast here.